router/router.go

```diff
@@ -2136,6 +2130,28 @@ func (rt *HandleT) readAndProcess() int {
 			drainedJobsStat.Count(destDrainStat.Count)
 			metric.DecreasePendingEvents("rt", destDrainStat.Workspace, rt.destName, float64(drainStatsbyDest[destID].Count))
 		}
+
+		if rt.guaranteeUserEventOrder {
+			for _, drainedJob := range drainJobList {
+				// cleanup failedJobIDMap for drained jobs
+				drainedUserID := drainedJob.UserID
+				partition := rt.getWorkerPartition(drainedUserID)
+				drainWorker := rt.workers[partition]
+
+				drainWorker.failedJobIDMutex.RLock()
+				lastJobID, ok := drainWorker.failedJobIDMap[drainedUserID]
+				drainWorker.failedJobIDMutex.RUnlock()
+
+				if ok && lastJobID == drainedJob.JobID {
+
+					drainWorker.failedJobIDMutex.Lock()
+					delete(drainWorker.failedJobIDMap, drainedUserID)
+					drainWorker.failedJobIDMutex.Unlock()
+
+				}
```

Comment:
There is no guarantee that between:
```go
drainWorker.failedJobIDMutex.RUnlock()
```
and
```go
drainWorker.failedJobIDMutex.Lock()
```

A different goroutine can modify the content of `drainWorker.failedJobIDMap[drainedUserID]` 

Maybe it is prevented by another mechanism and/or the probability is small.

However, given that drains are infrequent and the duration of the lock very small, I suggest taking the safest way here:

```suggestion
			   drainWorker.failedJobIDMutex.Lock()
				lastJobID, ok := drainWorker.failedJobIDMap[drainedUserID]

				if ok && lastJobID == drainedJob.JobID {
					delete(drainWorker.failedJobIDMap, drainedUserID)
				}
			    drainWorker.failedJobIDMutex.Unlock()

```

Reply:
> Maybe it is prevented by another mechanism and/or the probability is small.

Yes, by design, `readAndProcess` is the only one removing `userID`s from the `failedJobIDMap` and it does so under the assumption that there is no other outstanding job already in the pipeline for the same `userID`, otherwise ordering guarantees would be violated.

Nevertheless, the performance penalty of what you are proposing is minimal and promotes the correct transitioning pattern from a read lock to a write lock!